src/compiler/transformers/destructuring.ts

```diff
@@ -633,27 +384,84 @@ namespace ts {
-     * @param visitor An optional callback used to visit the value.
+     * @param location The location to use for source maps and comments.
      */
     function ensureIdentifier(
+        host: FlattenHost,
         value: Expression,
         reuseIdentifierExpressions: boolean,
-        location: TextRange,
-        emitTempVariableAssignment: (value: Expression, location: TextRange) => Identifier,
-        visitor?: (node: Node) => VisitResult<Node>) {
+        location: TextRange) {
 
         if (isIdentifier(value) && reuseIdentifierExpressions) {
             return value;
         }
         else {
-            if (visitor) {
-                value = visitNode(value, visitor, isExpression);
+            const temp = createTempVariable(/*recordTempVariable*/ undefined);
+            if (host.recordTempVariablesInLine) {
+                host.emitBindingOrAssignment(temp, value, location, /*original*/ undefined);
+            }
+            else {
+                host.context.hoistVariableDeclaration(temp);
+                host.emitExpression(createAssignment(temp, value, location));
             }
+            return temp;
+        }
+    }
+
+    function makeArrayBindingPattern(elements: BindingOrAssignmentElement[]) {
+        Debug.assertEachNode(elements, isArrayBindingElement);
+        return createArrayBindingPattern(<ArrayBindingElement[]>elements);
+    }
 
-            return emitTempVariableAssignment(value, location);
+    function makeArrayAssignmentPattern(elements: BindingOrAssignmentElement[]) {
+        return createArrayLiteral(map(elements, convertToArrayAssignmentElement));
+    }
+
+    function makeObjectBindingPattern(elements: BindingOrAssignmentElement[]) {
+        Debug.assertEachNode(elements, isBindingElement);
+        return createObjectBindingPattern(<BindingElement[]>elements);
+    }
+
+    function makeObjectAssignmentPattern(elements: BindingOrAssignmentElement[]) {
+        return createObjectLiteral(map(elements, convertToObjectAssignmentElement));
+    }
+
+    function makeBindingElement(name: Identifier) {
+        return createBindingElement(/*propertyName*/ undefined, /*dotDotDotToken*/ undefined, name);
+    }
+
+    function makeAssignmentElement(name: Identifier) {
+        return name;
+    }
+
+    /** Given value: o, propName: p, pattern: { a, b, ...p } from the original statement
+     * `{ a, b, ...p } = o`, create `p = __rest(o, ["a", "b"]);`*/
+    function createRestCall(value: Expression, elements: BindingOrAssignmentElement[], computedTempVariables: Expression[], location: TextRange): Expression {
+        const propertyNames: Expression[] = [];
+        for (let i = 0; i < elements.length - 1; i++) {
+            const propertyName = getPropertyNameOfBindingOrAssignmentElement(elements[i]);
+            if (propertyName) {
+                if (isComputedPropertyName(propertyName)) {
+                    // get the temp name and put that in there instead, like `_tmp + ""`

```

Comment:
can you update this to just give the emit: `typeof _tmp === "symbol" ? _tmp : _tmp + ""` ? The comment I wrote is wordy and out of date to boot.